# Lesson 2

In this lesson you'll get introduced to Pandas, a very powerful library for data science. It is among the most used tools in data science, which is why you'll use it this semester.

## Jupyter notebook

Please note that I'll be using Jupyter Notebooks to teach you. All of this code works perfectly fine if you export it to a Python script (xxx.py file). In fact, a Jupyter Notebook itself is one big JSON file with both code and output defined. You can do whatever you want with that JSON structure. There's a few plugins that allow you to separate code and output. There's even a plugin that allows you to write your scripts in a separate file and then run them as a code cell.

## This lesson

When this lesson is over, you can write programs that read datasets, perform some filterings and groupings on them, and gain true understanding of your data as a result!

## The dataset

We'll use Kaggle's [Video Games Sales Dataset](https://www.kaggle.com/sidtwr/videogames-sales-dataset), especially the [.csv file for all video games](https://www.kaggle.com/sidtwr/videogames-sales-dataset?select=Video_Games_Sales_as_at_22_Dec_2016.csv). It's up for grabs at [Learn.HZ](https://learn.hz.nl) as well.

Please download the file and put it in a directory next to this notebook. Other directories are fine as well, but you'll have to debug your own paths since Python can be quite unintuitive about them if you don't know what you're doing.

In [1]:
try:
    with open("./Video_Games_sales_as_at_22_Dec_2016.csv") as f:
        print("Let's look at the homework from previous lesson first!")
except IOError:
        print("File is not there! :(")

Let's look at the homework from previous lesson first!


## Don't forget them imports!

Note that we abbreviate the import. This isn't necessary but pretty much everyone uses these abbreviations. So might as well get used to it as a best practice.

In [2]:
import pandas as pd

## Dataframe

Pandas uses *Dataframes*, something directly leveraged from the R language. A dataframe sounds nice, and while it is, it is essentially a table. Let's read it in and take a look at it. We use the `read_csv` function because our dataset is in .csv format. Note that the `read_csv` function contains a lot of parameters, for example for delimiters. You will probably need to use those in your own datasets.

In [3]:
df = pd.read_csv("./Video_Games_sales_as_at_22_Dec_2016.csv", delimiter=";")

Normally, you would need to write some elegant code to print complicated datatypes to your screen. But in Jupyter, you can just create a cell that contains only the variable name and then it automatically prints it for you.

In [4]:
df

Name Platform  Year_of_Release         Genre  \
0                         Wii Sports      Wii           2006.0        Sports   
1                  Super Mario Bros.      NES           1985.0      Platform   
2                     Mario Kart Wii      Wii           2008.0        Racing   
3                  Wii Sports Resort      Wii           2009.0        Sports   
4           Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing   
...                              ...      ...              ...           ...   
16714  Samurai Warriors: Sanada Maru      PS3           2016.0        Action   
16715               LMA Manager 2007     X360           2006.0        Sports   
16716        Haitaka no Psychedelica      PSV           2016.0     Adventure   
16717               Spirits & Spells      GBA           2003.0      Platform   
16718            Winning Post 8 2016      PSV           2016.0    Simulation   

          Publisher  NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  \
0          Nintendo     41.36     28.96      3.77         8.45         82.53   
1          Nintendo     29.08      3.58      6.81         0.77         40.24   
2          Nintendo     15.68     12.76      3.79         3.29         35.52   
3          Nintendo     15.61     10.93      3.28         2.95         32.77   
4          Nintendo     11.27      8.89     10.22         1.00         31.37   
...             ...       ...       ...       ...          ...           ...   
16714    Tecmo Koei      0.00      0.00      0.01         0.00          0.01   
16715   Codemasters      0.00      0.01      0.00         0.00          0.01   
16716  Idea Factory      0.00      0.00      0.01         0.00          0.01   
16717       Wanadoo      0.01      0.00      0.00         0.00          0.01   
16718    Tecmo Koei      0.00      0.00      0.01         0.00          0.01   

       Critic_Score  Critic_Count  User_Score  User_Count Developer Rating  
0              76.0          51.0         8.0       322.0  Nintendo      E  
1               NaN           NaN         NaN         NaN       NaN    NaN  
2              82.0          73.0         8.3       709.0  Nintendo      E  
3              80.0          73.0         8.0       192.0  Nintendo      E  
4               NaN           NaN         NaN         NaN       NaN    NaN  
...             ...           ...         ...         ...       ...    ...  
16714           NaN           NaN         NaN         NaN       NaN    NaN  
16715           NaN           NaN         NaN         NaN       NaN    NaN  
16716           NaN           NaN         NaN         NaN       NaN    NaN  
16717           NaN           NaN         NaN         NaN       NaN    NaN  
16718           NaN           NaN         NaN         NaN       NaN    NaN  

[16719 rows x 16 columns]

**Assignment (10 min)**: Answer the following questions

1. How many rows does the dataframe contain? **16719**
2. How many columns does the dataframe contain? **16**
3. How many cells are there in the dataframe? **267.504**
4. Does every cell actually contain data? **No, there are quite some `NaN` cells**
5. Are there columns you don't understand? **Rating? This is the ESRB rating, the US' Kijkwijzer**
6. Are there columns with data that probably mean something else than what is literally there? **Sales. A game like Wii Sports will probably have been  sold 41 million times instead of just 41**.
7. What is the unit of analysis? **The entity you are going to analyze. Throughout this lesson, we will change this unit of analysis, pay close attention!**
8. What is the unit of observation? **A unit of observation is literally the entity you are "observing". In this case, a game.**
9. Is there weird data in any of the cells for so far as you can see? (weird = i.e. 8 billion sales, or -100) **A couple of Sales cells are 0.00?**

**Assignment (5 min)**: You are now in the Business Understanding and Data Understanding phase of the CRISP-DM cycle. Write down at least three questions to ask your imaginary client based on your observations of the dataframe above.

## Using Pandas

Let's take a closer look at the data itself. I've identified a couple of curious things in it for you, to save time. After all, the goal of this lecture is to learn you the tools do start doing it yourself! Keep in mind that you would normally investigate most columns and rows. Don't underestimate how many hours it can take to get a good grasp of what the data is telling you. Spend time digging through it, very thoroughly! At least a couple of hours per week for the first quartile.

### Columns

It is often useful to just count values in column, or check how many unique values there are. Selecting columns in a dataframe is like working with a dictionary: `variable_name["column_name"]`. The `len()` function is a built-in Python function that checks how long (length) a list, dictionary or tuple is.

In [ ]:
amount_of_names = len(df["Name"])
unique_names = len(df["Name"].unique())

print("Contains", amount_of_names, "names")
print("Contains", unique_names, "unique names")

Hmm, there are approximately 5000 names that aren't unique. Let's find those!

In [ ]:
df["Name"].value_counts()

**Assignment (3 min)**: Think of reasons why some games come up more than once in the dataset. Use your domain knowledge of gaming, or browse the internet to validate your answer

We could have, of course, just checked the data more closely to see what was going on. We will use the `duplicated` function in Pandas for this. Fortunately, we can see that our assumption is correct by comparing row 1 to row 5: GTA V was released on multiple platforms. Also note the quite awkward code to create the dataframe. We'll break down what is going on below.

In [ ]:
df[df.duplicated(["Name"])]

Let's break down what the heck that awkward code means! `df.duplicated()` is a function that takes a column in the dataframe. Since we're calling the function **on** a dataframe, we don't have to specify the variable name of our dataframe when putting the column name in square brackets. The line below returns a Pandas Series object -- essentially an array -- with the index of every row in the dataframe. Behind that index is a boolean: does that row contain a duplicated name or not? This technique is also called *[masking](https://pythonhealthcare.org/2018/04/07/30-using-masks-to-filter-data-and-perform-search-and-replace-in-numpy-and-pandas/)*.

In [ ]:
mask = df.duplicated(["Name"])
print(mask)

If we now put this *mask* inside square brackets of the original dataframe again, Pandas understands that we only want to see the rows that contain a True value in the mask. It's a bit of magic that takes some getting used to, but is very powerful indeed. Note that the results below are exactly the same as above.

In [ ]:
df[mask]

Since we're at least a little bit geeky, and I have some serious nostalgia for Need for Speed: Most Wanted, let's find out on which platforms it was released! We'll use a *mask* again. Note that this time we just write a boolean expression instead of using a function like `duplicated()`.

In [ ]:
df[df["Name"] == "Need for Speed: Most Wanted"]

Jupyter is kind enough to show the whole dataframe, but let's just practice taking out only the Platform values. Your datasets may be a lot larger than this one, and Jupyter will the abbreviate your dataframes when you print them...

In [ ]:
df[df["Name"] == "Need for Speed: Most Wanted"]["Platform"].unique()

## Intermission

Note that the line above is very much "data scientist code". It works, does what you want and does that well. And if you come back to that code in 6 months, you will absolutely have no clue what's going on. Since I'm a software engineer at heart, I will always encourage you to write a bit *better* code once you have done your preliminary digging around. Something like the cell below is, in my opinion, a better understandable version than that insane one-liner above. We will not grade the code you ~~will vomit at the screen~~ write, but we will grade your understanding of the code. Please keep that in mind every time you write code that you probably will not change for a while.

In [ ]:
nfs_mask = df["Name"] == "Need for Speed: Most Wanted"
nfs_df = df[nfs_mask]
unique_platforms_for_nfs = nfs_df["Platform"].unique()
print(unique_platforms_for_nfs)

## Your turn!

**Assignment (15 min)**: Use your gained knowledge on the `Year_of_Release` column to identify 4 very strange cases. Then print a dataframe that only contains the rows of the strange cases. Don't know what you're looking for? Take a good look at the filename of your dataset, and then take a good look at all the unique values in the column.

In [ ]:
mask = df["Year_of_Release"] > 2016
df[mask]

**Assignment (5 min)**: Revisit the list of questions from earlier. Have the answers for any of them changed?

## Grouping and sorting

Sometimes it can be very useful to group some data together. The `df["column_name"].value_counts()` function you have used, is actually a shortcut function to writing `df.groupby(column_name)["column_name"].count()`. The first one is... a bit easier though, right? Let's see what grouping can do for us.

In [ ]:
df.groupby("Name")["EU_Sales"].sum().sort_values()

Wooooaaaah, not so fast. What is going on?

First, create a dataframe that is grouped by name. For example, all Need for Speed: Most Wanted games will be thrown onto one big heap.

In [ ]:
grouped_df =  df.groupby("Name")
grouped_df

Now let's check some of the columns, pick anyone you find interesting! It returns a very vague object again. This is because some of the rows (in this case, all the duplicated Name rows) have been thrown onto one big heap. Pandas has no clue what platform to show, or what Critic_Score or what EU_Sales.

In [ ]:
grouped_df["EU_Sales"]

But we can use that vague object to do some cool stuff on. For example, let's take the sum of all the sales for a game in the EU. This return a Pandas Series object (an array) again.

In [ ]:
grouped_df["EU_Sales"].sum()

**Assignment (5 min)**: Take a moment to step back and check the three code cells above. Why does this summing of a vague `groupby` object work? Feel free to draw it out, use the internet, come up with a simpler, imaginary dataset, etc.

This series object is usable, but it can be improved. Let's sort it so that we can see what games have sold very well, and which ones haven't sold at all.

In [ ]:
grouped_df["EU_Sales"].sum().sort_values()

## Slicing and dicing

Lists, arrays and other iterable objects (such as Pandas Series) can be sliced in Python. It looks strange but works very intuitively. Very often, you will use the `:` operator. You can read that as "everything before/after".

In [ ]:
example_list = [10, 20, 30, 40, 50, 60, 70, 80, 90, 100]
print(example_list)

In [ ]:
# Print all the items from index 5 onwards. Read it as: "print index 5 and everything after"
print(example_list[5:])

# Print all the items up to index 5. Read it as: "print everything before index 5"
print(example_list[:5])

#Print all the items between index 2 and 7. Read it as "print index 2 and everything after, but before index 7"
print(example_list[2:7])

Python also support negative indexes. When you use these, you basically start counting at the end of the list instead of the beginning.

In [ ]:
# Print the last item in the list
print(example_list[-1])

# Print the last three items in the list. Read it as "print index 3 and everything after while starting at the end of the list"
print(example_list[-3:])

# Print all the items up to the index. Read it as "print everything before index 3 while starting at the end of the list"
print(example_list[:-3])

## Putting it all together

Let's take the top 5 of the sorted EU sales we calculated earlier.

In [ ]:
grouped_df["EU_Sales"].sum().sort_values()[-5:]

Turns out people like a Wii, soccer and shooting each other, huh?

**Assignment (20 to 30 minutes)**: Find the the top 5 games that are scored 80 or higher by at least 25 critics, sorted on global sales. You **do not** have to take measures to prevent games that were re-released (like NFS: Most Wanted) from polluting the data. For this exercise, you may just mash them together :)

## That's all, folks!

Pandas is a very powerful library and can do a lot more than we've seen today. On Learn is a cheat sheet with some common commands for use cases you will probably all run into.

Feel like you need some extra exercise? The Kaggle tutorial on Pandas covers most of what you will need this semester and takes about 1.5-2 hours to complete: https://www.kaggle.com/learn/pandas